webapp/channels/src/utils/burn_on_read_expiration_scheduler.ts

```diff
@@ -0,0 +1,316 @@
+                this.checkAndExpirePosts();
+
+                // Note: checkAndExpirePosts() calls recomputeSchedule() if needed
+            }, delay);
+        } else {
+            // Long delay: use polling to handle sleep/throttling
+            // Check immediately in case we woke from sleep
+            this.checkAndExpirePosts();
+
+            // Then poll periodically
+            this.pollingIntervalId = setInterval(() => {
+                this.checkAndExpirePosts();
+            }, this.pollingInterval);
+        }
+    }
+
+    /**
+     * Get the next post that will expire
+     */
+    private getNextExpiring(): {postId: string | null; expireAt: number} {
+        let nextPostId: string | null = null;
+        let nextExpiration = Infinity;
+
+        for (const [postId, expireAt] of this.postExpirations.entries()) {
+            if (expireAt < nextExpiration) {
+                nextExpiration = expireAt;
+                nextPostId = postId;
+            }
+        }
+
+        return {postId: nextPostId, expireAt: nextExpiration};
+    }
+
+    /**
+     * Check all posts and expire any that have passed their expiration time + grace period
+     */
+    private checkAndExpirePosts(): void {
+        const now = Date.now();
+        const expiredPosts: string[] = [];
+
+        // Find all expired posts (including grace period in the check)
+        // This is where the grace period should be applied, not just in registerPost
+        for (const [postId, expireAt] of this.postExpirations.entries()) {
+            if (expireAt + this.gracePeriodMs <= now) {
+                expiredPosts.push(postId);
+            }
+        }
+
+        // Batch delete from tracking (avoids mutation during iteration)
+        for (const postId of expiredPosts) {
+            this.postExpirations.delete(postId);
+        }
+
+        // Dispatch batch expiration if dispatch is available
+        if (expiredPosts.length > 0 && this.dispatch) {
+            // Dispatch each expiration (in future, consider batched action to reduce Redux thrash)
+            for (const postId of expiredPosts) {
+                try {
+                    this.dispatch(handlePostExpired(postId));
+                } catch (error) {
+                    // Silently fail - errors are logged by Redux middleware
+                }
+            }
+        }
+
+        // Reschedule if we expired any posts
+        // - If we expired posts and more remain: schedule the next one
+        // - If we expired posts and none remain: recomputeSchedule() exits early
+        if (expiredPosts.length > 0) {
+            this.recomputeSchedule();
+        }
+    }
+
+    /**
+     * Handle visibility change - check expirations when tab becomes visible
+     */
+    private handleVisibilityChange = (): void => {
+        if (!document.hidden) {
+            // Tab became visible, check for expired posts
+            this.checkAndExpirePosts();
```

Comment:
This will be called from call to `this.recomputeSchedule()` in the `if` block below, so is this call redundant?